### PR TITLE
feat: support uploading active log files

### PR DESCRIPTION
**Pull request by dgggIoT**
**Issue #, if available:**
#51 
**Description of changes:**
add hash generation and add unit tests
**Why is this change necessary:**

**How was this change tested:**
Unit test
**Any additional information or context required to review the change:**
This is not finished yet. Please do not merge. 
**Checklist:**
 - [ ] Updated the README if applicable
 - [x ] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessor.java [hunk 1]
```diff
@@ -222,6 +223,9 @@ private boolean processLogLine(AtomicInteger totalBytesRead,
         CloudWatchAttemptLogInformation attemptLogInformation;
         Optional<GreengrassLogMessage> logMessage = tryGetStructuredLogMessage(dataStr);
         Pair<Boolean, AtomicInteger> addEventResult;
+        // if the message is in JSON format, directly grab the message timestamp as logStreamName
+        // otherwise, try to find if any timestamp matched in text. If not, set the local (where LM is) time as the
+        // logStreamName
```
Comment (dgggIoT) [inline src/main/java/com/aws/greengrass/logmanager/CloudWatchAttemptLogsProcessor.java:228]: These are not related to the change. I just add some comments.

File: src/test/java/com/aws/greengrass/logmanager/util/FileHashGeneratorTest.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        try (OutputStream fileOutputStream = Files.newOutputStream(file.toPath())) {
+            for (int i = 0; i < DEFAULT_LINES_FOR_DIGEST_NUM; i++) {
+                fileOutputStream.write("line".getBytes(StandardCharsets.UTF_8));
+                fileOutputStream.write(String.valueOf(i + 1).getBytes(StandardCharsets.UTF_8));
+                fileOutputStream.write("\n".getBytes(StandardCharsets.UTF_8));
+            }
+        }
+        try {
+            Optional fileHash = FileHashGenerator.calculateHashForLogFile(file, DEFAULT_LINES_FOR_DIGEST_NUM);
+            assertFalse(fileHash.isPresent());
```
Comment (Nelsonochoam) [inline src/test/java/com/aws/greengrass/logmanager/util/FileHashGeneratorTest.java:72]: We could probably use a better assertion. Is there a way we could check that the hash is what we expect?. What is the `Optional` holding should it be `Optional<String>`?

File: src/test/java/com/aws/greengrass/logmanager/util/FileHashGeneratorTest.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        finally {
+            assertTrue(file.delete());
+        }
+    }
+
+    @Test
+    void GIVEN_log_file_with_equal_to_target_lines_WHEN_calculate_file_hash_THEN_we_get_null() throws IOException {
+        int DEFAULT_LINES_FOR_DIGEST_NUM = 1;
+        File file = new File(directoryPath.resolve("greengrass_test.log").toUri());
+        try (OutputStream fileOutputStream = Files.newOutputStream(file.toPath())) {
```
Comment (Nelsonochoam) [inline src/test/java/com/aws/greengrass/logmanager/util/FileHashGeneratorTest.java:63]: We could extract this into a method something called `givenALogFile`

File: src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        
+    }
+
+    /**
+     * calculate the hash for the log file, which takes and digests first some lines (default is 1) in the file.
+     * If the content of file is insufficient, the hash is null.
+     * @param logFile
+     * @return the nullable calculated hash for active file
+     */
+    public static Optional<String> calculateHashForLogFile(File logFile, int DEFAULT_LINES_FOR_DIGEST_NUM) {
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java:30]: use camelCase for argument names if you want a "default" values use method overloading. For instance

```java
calculateHashForLogFile(File logFile) {
      calculateHashForLogFile(logFile, DEFAULT_LINES_FOR_DIGEST_NUM);
}

calculateHashForLogFile(File logFile, int digestLinesLength) {
      calculateHashForLogFile(logFile, DEFAULT_LINES_FOR_DIGEST_NUM);
}

```

File: src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import java.nio.file.Files;
+import java.security.NoSuchAlgorithmException;
+import java.util.Optional;
+
+import static com.aws.greengrass.util.Digest.calculate;
+
+public final class FileHashGenerator {
+    private static final Logger logger = LogManager.getLogger(LogManagerService.class);
+    // private constructor to make checkstyle happy
+    private FileHashGenerator() {
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java:20]: Should not complain if you remove the entire constructor?

File: src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    /**
+     * calculate the hash for the log file, which takes and digests first some lines (default is 1) in the file.
+     * If the content of file is insufficient, the hash is null.
+     * @param logFile
+     * @return the nullable calculated hash for active file
+     */
+    public static Optional<String> calculateHashForLogFile(File logFile, int DEFAULT_LINES_FOR_DIGEST_NUM) {
+        Optional fileHash = Optional.ofNullable(null);
+        // empty file, then do not read
+        if (!logFile.isFile())
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java:33]: This does not check if the file is empty. It checks if it is a directory or a file.

File: src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            return fileHash;
+        try (BufferedReader r = Files.newBufferedReader(logFile.toPath(), StandardCharsets.UTF_8)) {
+            StringBuilder data = new StringBuilder();
+            int linesReadForDigest = 0;
+            String oneLine;
+            // read target number of lines; read one more line to prevent incomplete last line
+            // Therefore, there must be DEFAULT_LINES_FOR_DIGEST_NUM + 1 lines in each file.
+            while (linesReadForDigest <= DEFAULT_LINES_FOR_DIGEST_NUM) {
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java:41]: This could be split into 2 methods:

1) Get lines from a file
2) Hash the content of the lines

File: src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+     * @param logFile
+     * @return the nullable calculated hash for active file
+     */
+    public static Optional<String> calculateHashForLogFile(File logFile, int DEFAULT_LINES_FOR_DIGEST_NUM) {
+        Optional fileHash = Optional.ofNullable(null);
+        // empty file, then do not read
+        if (!logFile.isFile())
+            return fileHash;
+        try (BufferedReader r = Files.newBufferedReader(logFile.toPath(), StandardCharsets.UTF_8)) {
+            StringBuilder data = new StringBuilder();
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java:36]: Maybe call it `logLines`?

File: src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java (hunk 7 above)
Comment (Nelsonochoam) [reply src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java:41]: Also this could be improved a bit

This method would just read lines 
```java
List<String> linesNeeded = new ArrayList();

try (BufferedReader reader = Files.newBufferedReader(logFile.toPath(), StandardCharsets.UTF_8)) {
     String line = reader.readLine();

     if (line == null) {
        return linesNeeded;
     }

     linesNeeded.append(line);

      while (line != null && linesNeeded.size() < DEFAULT_LINES_FOR_DIGEST_NUM) {
          line = reader.readLine();
          if (line != null) {
              linesNeeded.append(line);
          }
     }
}

return linesNeeded;
```

Then `calculateHashForLogFile` calls it gets the lines and runs the logic to check if it can return a valid hash

File: src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.security.NoSuchAlgorithmException;
+import java.util.Optional;
+
+import static com.aws.greengrass.util.Digest.calculate;
+
+public final class FileHashGenerator {
```
Comment (Nelsonochoam) [inline src/main/java/com/aws/greengrass/logmanager/util/FileHashGenerator.java:17]: It would be nice if we had a wrapper around the File class instead of having this `Generator` class which doesn't have much meaning. Maybe we have

```java
public class LogFile extends File {
     //... call super and so on

     getHash() <- your method

}

```
